teststand/teststand.1
.\"
.\" Copyright © 2018 Bdale Garbee <bdale@gag.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH ALTOSUI 1 "teststand" ""
.SH NAME
teststand \- Wireless Test Stand control program
.SH SYNOPSIS
.B "teststand"
.SH DESCRIPTION
.I teststand
connects to a TeleDongle or TeleBT device through a USB serial device.
It provides a menu-oriented
user interface to initiate, monitor, record and review rocket motor 
test stand data.
.SH USAGE
With a TeleDongle or TeleBT attached, teststand makes it possible to
control a remote Altus Metrum wireless test stand board.  
.SH FILES
All downloaded test stand data files are recorded into a user-specified 
directory (default ~/TeleMetrum). Files are named using the current date, 
the serial number of the reported device, the test number recorded in 
the data and '.eeprom' for eeprom data.
.SH AUTHOR
Bdale Garbee


ChangeLog
commit dbb78c8222c45f4430601deee0194b0c9dc2e79a
Merge: fe38c225 87aab995
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Dec 21 19:05:46 2017 -0700

    Merge branch 'master' into branch-1.8

commit 87aab99521dc44d1d29fbb0b7f227f868f074836
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Dec 21 17:37:10 2017 -0800

    Bump for version 1.8.4
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 4bcdc106df2c5e8572570e57b4d97121df94799a
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Dec 21 17:36:24 2017 -0800

    Doc updates for version 1.8.4
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 46304aa257635d14afc4d8567eedba0f93a5742f
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 19 16:57:33 2017 -0800

    altos/micropeak: Remove all compiler results
    
    Not just the current version.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 99299986e194337b05ee81cfb7c4aa1cb9e9a74e
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 19 16:56:33 2017 -0800

    altos/attiny: Add ADC implementation
    
    It's primitive, but might serve to read ADC values. Untested.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 342132a8869d530b6893bb84becf03cb30490600
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 19 16:21:23 2017 -0800

    ao-bringup: turnon_easymini was left with dfu_util disabled
    
    This was presumably changed for some test and left in the wrong
    state.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 6d05747941b8e31afb1f8522ac0b8c1ad12aa90f
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 19 16:19:40 2017 -0800

    altos/easymini-v2.0: Adapt to final hardware pin assignment changes
    
    Beeper moved from PB0 to PB1 (Tim3 CH4)
    Drogue fire moved from PB6 to PB3
    Bootloader moved from PB1 to PB6
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 9826845f952abe898f029e31cc0f7080708e2eae
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 19 13:09:24 2017 -0800

    altos/lambdakey-v1.0: Add back and/or macros
    
    With scheme shrinking a bit, there's now space for these useful macros.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit ed1f7b79abc7400a54b35fbf62c9db6855f9129a
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 19 12:39:20 2017 -0800

    altos/scheme: Replace per-type indexed stash with poly stash heap
    
    Instead of having a random set of stash arrays with explicit indices
    used by callers, just have a general heap. Less error prone, and less code.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 71fb79492cb955af4bd52e79f1fa69d17e084dbc
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 19 12:16:24 2017 -0800

    altos/scheme: Replace memory pool macros with inlines
    
    AO_SCHEME_IS_CONST -> ao_scheme_is_const_addr
    AO_SCHEME_IS_POOL -> ao_scheme_is_pool_addr
    
    Provides better typechecking and avoids confusion with
    ao_scheme_is_const inline (which takes an ao_poly instead of a void *)
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 34f998d147d08e966daad1ab76c40906018d3d8d
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 19 11:51:33 2017 -0800

    altos/scheme: AO_SCHEME_IS_CONS -> ao_scheme_is_cons
    
    This inline was already defined; just use it. Also, switch some places
    to use ao_scheme_is_pair instead as appropriate.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit fa6f4b331db9d37da6767005fd375b696485b46b
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 19 11:43:23 2017 -0800

    altos/scheme: ao_scheme__cons -> ao_scheme_cons
    
    Fix the double underscore in this name. Ick.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 00390fb09f47654905824af671b966ffca0a38b3
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 19 11:40:08 2017 -0800

    altos/scheme: Don't optimize ao_scheme_make_const
    
    Performance isn't interesting, and it's nice to have a bare system
    ready for debugging.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit ba472dda57e134fe0f0e4a571a6d0c1e5a1ea6eb
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 19 11:39:39 2017 -0800

    altos/scheme: Clean up test CFLAGS
    
    make it easy to switch between debug and optimized builds.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 53b99e0419cb44c7983e41026bf0430deae58940
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 19 11:37:33 2017 -0800

    altos/scheme: (define (foo . bar)) has a pair, not list as card
    
    When defining a lambda with varargs, the args are not a list as the
    final element is not a pair or nil. Use pair? instead of list? to
    detect this form correctly.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit fbe5dc9f215e7014aa8f9d325c1fba939816be03
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 19 11:35:09 2017 -0800

    altos/scheme: apply also needs to not free value list on lambdas
    
    When apply is invoked on any function, the cons in the argument list
    cannot be immediately freed as they have been passed to the
    function. That applies to both built-ins as well as lambdas; this
    patch removes the special ao_scheme_skip_cons_free global and just
    marks the stack in both cases.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 431165e5fa72ba6dffd477de32960745cdec332c
Author: Keith Packard <keithp@keithp.com>
Date:   Tue Dec 19 11:33:36 2017 -0800

    altos/scheme: Rework display/write code
    
    Unify output functions and add bool to switch between write and
    display mode. Make that only affect strings (as per r⁷rs).
    
    Use print recursion detection in frame and stack code, eliminating
    PRINT flags in type field.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>


teststand/TestStand.java
/*
 * Copyright © 2010 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

package teststand;

import java.awt.*;
import java.awt.event.*;
import javax.swing.*;
import java.io.*;
import java.util.concurrent.*;
import org.altusmetrum.altoslib_12.*;
import org.altusmetrum.altosuilib_12.*;

public class TestStand extends AltosUIFrame implements AltosEepromGrapher {
	public AltosVoice voice = new AltosVoice();

	public static boolean load_library(Frame frame) {
		if (!Altos.load_library()) {
			JOptionPane.showMessageDialog(frame,
						      String.format("No AltOS library in \"%s\"",
								    System.getProperty("java.library.path","<undefined>")),
						      "Cannot load device access library",
						      JOptionPane.ERROR_MESSAGE);
			return false;
		}
		return true;
	}

	void telemetry_window(AltosDevice device) {
		try {
			AltosFlightReader reader = new AltosTelemetryReader(new AltosSerial(device));
			if (reader != null)
				new AltosFlightUI(voice, reader, device.getSerial());
		} catch (FileNotFoundException ee) {
			JOptionPane.showMessageDialog(TestStand.this,
						      ee.getMessage(),
						      String.format ("Cannot open %s", device.toShortString()),
						      JOptionPane.ERROR_MESSAGE);
		} catch (AltosSerialInUseException si) {
			JOptionPane.showMessageDialog(TestStand.this,
						      String.format("Device \"%s\" already in use",
								    device.toShortString()),
						      "Device in use",
						      JOptionPane.ERROR_MESSAGE);
		} catch (IOException ee) {
			JOptionPane.showMessageDialog(TestStand.this,
						      String.format ("Unknown I/O error on %s", device.toShortString()),
						      "Unknown I/O error",
						      JOptionPane.ERROR_MESSAGE);
		} catch (TimeoutException te) {
			JOptionPane.showMessageDialog(this,
						      String.format ("Timeout on %s", device.toShortString()),
						      "Timeout error",
						      JOptionPane.ERROR_MESSAGE);
		} catch (InterruptedException ie) {
			JOptionPane.showMessageDialog(this,
						      String.format("Interrupted %s", device.toShortString()),
						      "Interrupted exception",
						      JOptionPane.ERROR_MESSAGE);
		}
	}

	public void scan_device_selected(AltosDevice device) {
		telemetry_window(device);
	}

	Container	pane;
	GridBagLayout	gridbag;

	JButton addButton(int x, int y, String label) {
		GridBagConstraints	c;
		JButton			b;

		c = new GridBagConstraints();
		c.gridx = x; c.gridy = y;
		c.fill = GridBagConstraints.BOTH;
		c.weightx = 1;
		c.weighty = 1;
		b = new JButton(label);

		//Dimension ps = b.getPreferredSize();

		gridbag.setConstraints(b, c);
		add(b, c);
		return b;
	}

	/* OSXAdapter interfaces */
	public void macosx_file_handler(String path) {
		process_graph(new File(path));
	}

	public void macosx_quit_handler() {
		System.exit(0);
	}

	public void macosx_preferences_handler() {
		ConfigureTestStand();
	}

	public TestStand() {

		load_library(null);

		register_for_macosx_events();

		AltosUIPreferences.set_component(this);

		pane = getContentPane();
		gridbag = new GridBagLayout();
		pane.setLayout(gridbag);

		JButton	b;

		b = addButton(0, 0, "Monitor Flight");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						ConnectToDevice();
					}
				});
		b.setToolTipText("Connect to TeleDongle and monitor telemetry");
		b = addButton(1, 0, "Save Data");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						SaveFlightData();
					}
				});
		b.setToolTipText("Download and/or delete flight data from an altimeter");
		b = addButton(3, 0, "Graph Data");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						GraphData();
					}
				});
		b.setToolTipText("Present flight data in a graph and table of statistics");
		b = addButton(4, 0, "Export Data");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						ExportData();
					}
				});
		b.setToolTipText("Convert flight data for a spreadsheet or GoogleEarth");
		b = addButton(0, 1, "Configure TeleFire");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						ConfigureTeleMetrum();
					}
				});
		b.setToolTipText("Set flight, storage and communication parameters");
		b = addButton(1, 1, "Configure TestStand");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					ConfigureTestStand();
				}
			});
		b.setToolTipText("Global TestStand settings");

		b = addButton(2, 1, "Configure Ground Station");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					ConfigureTeleDongle();
				}
			});

		b = addButton(3, 1, "Flash Image");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					FlashImage();
				}
			});
		b.setToolTipText("Replace the firmware in any AltusMetrum product");

		b = addButton(0, 2, "Scan Channels");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					ScanChannels();
				}
			});
		b.setToolTipText("Find what channel an altimeter is sending telemetry on");

		b = addButton(4, 2, "Quit");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					System.exit(0);
				}
			});
		b.setToolTipText("Close all active windows and terminate TestStand");

		setTitle("Altus Metrum Test Stand");

		pane.doLayout();
		pane.validate();

		doLayout();
		validate();

		setDefaultCloseOperation(JFrame.DO_NOTHING_ON_CLOSE);
		addWindowListener(new WindowAdapter() {
			@Override
			public void windowClosing(WindowEvent e) {
				System.exit(0);
			}
		});

		setLocationByPlatform(false);

		/* Insets aren't set before the window is visible */
		setVisible(true);
	}

	private void ConnectToDevice() {
		AltosDevice	device = AltosDeviceUIDialog.show(TestStand.this,
								Altos.product_basestation);

		if (device != null)
			telemetry_window(device);
	}

	void ConfigureCallsign() {
		String	result;
		result = JOptionPane.showInputDialog(TestStand.this,
						     "Configure Callsign",
						     AltosUIPreferences.callsign());
		if (result != null)
			AltosUIPreferences.set_callsign(result);
	}

	void ConfigureTeleMetrum() {
		new AltosConfigFC(TestStand.this);
	}

	void ConfigureTeleDongle() {
		new AltosConfigTD(TestStand.this);
	}

	void FlashImage() {
		AltosFlashUI.show(TestStand.this);
	}

	void ScanChannels() {
		new AltosScanUI(TestStand.this, true);
	}

	/* Connect to TeleMetrum, either directly or through
	 * a TeleDongle over the packet link
	 */

	public void graph_flights(AltosEepromList flights) {
		for (AltosEepromLog flight : flights) {
			if (flight.graph_selected && flight.file != null) {
				process_graph(flight.file);
			}
		}
	}

	private void SaveFlightData() {
		new AltosEepromManage(this, this, AltosLib.product_any);
	}

	private static AltosFlightSeries make_series(AltosRecordSet set) {
		AltosFlightSeries series = new AltosFlightSeries(set.cal_data());
		set.capture_series(series);
		series.finish();
		return series;
	}

	/* Load a flight log file and write out a CSV file containing
	 * all of the data in standard units
	 */

	private void ExportData() {
		AltosDataChooser chooser;
		chooser = new AltosDataChooser(this);
		AltosRecordSet set = chooser.runDialog();
		if (set == null)
			return;
		AltosFlightSeries series = make_series(set);
		new AltosCSVUI(TestStand.this, series, chooser.file());
	}

	/* Load a flight log CSV file and display a pretty graph.
	 */

	private void GraphData() {
		AltosDataChooser chooser;
		chooser = new AltosDataChooser(this);
		AltosRecordSet set = chooser.runDialog();
		if (set == null)
			return;
		try {
			new AltosGraphUI(set, chooser.file());
		} catch (InterruptedException ie) {
		} catch (IOException ie) {
		}
	}

	private void ConfigureTestStand() {
		new AltosConfigureUI(TestStand.this, voice);
	}

	static AltosWriter open_csv(File file) {
		try {
			return new AltosCSV(file);
		} catch (FileNotFoundException fe) {
			System.out.printf("%s\n", fe.getMessage());
			return null;
		}
	}

	static AltosRecordSet record_set(File input) {
		try {
			return AltosLib.record_set(input);
		} catch (IOException ie) {
			String message = ie.getMessage();
			if (message == null)
				message = String.format("%s (I/O error)", input.toString());
			System.err.printf("%s: %s\n", input.toString(), message);
		}
		return null;
	}

	static final int process_none = 0;
	static final int process_csv = 1;
	static final int process_graph = 3;
	static final int process_summary = 4;
	static final int process_oneline = 5;

	static boolean process_csv(File input) {
		AltosRecordSet set = record_set(input);
		if (set == null)
			return false;

		File output = Altos.replace_extension(input,".csv");
		System.out.printf("Processing \"%s\" to \"%s\"\n", input, output);
		if (input.equals(output)) {
			System.out.printf("Not processing '%s'\n", input);
			return false;
		} else {
			AltosWriter writer = open_csv(output);
			if (writer == null)
				return false;
			AltosFlightSeries series = make_series(set);
			writer.write(series);
			writer.close();
		}
		return true;
	}

	static boolean process_graph(File file) {
		AltosRecordSet set = record_set(file);
		if (set == null)
			return false;
		try {
			new AltosGraphUI(set, file);
			return true;
		} catch (InterruptedException ie) {
		} catch (IOException ie) {
		}
		return false;
	}

	static boolean process_summary(File file) {
		AltosRecordSet set = record_set(file);
		if (set == null)
			return false;
		System.out.printf("%s:\n", file.toString());
		AltosFlightSeries series = make_series(set);
		TestStats stats = new TestStats(series);
		if (stats.serial != AltosLib.MISSING)
			System.out.printf("Serial:       %5d\n", stats.serial);
		if (stats.flight != AltosLib.MISSING)
			System.out.printf("Flight:       %5d\n", stats.flight);
		if (stats.year != AltosLib.MISSING)
			System.out.printf("Date:    %04d-%02d-%02d\n",
					  stats.year, stats.month, stats.day);
		if (stats.hour != AltosLib.MISSING)
			System.out.printf("Time:      %02d:%02d:%02d UTC\n",
					  stats.hour, stats.minute, stats.second);
		if (stats.max_height != AltosLib.MISSING)
			System.out.printf("Max height:  %6.0f m    %6.0f ft\n",
					  stats.max_height,
					  AltosConvert.meters_to_feet(stats.max_height));
		if (stats.max_speed != AltosLib.MISSING)
			System.out.printf("Max speed:   %6.0f m/s  %6.0f ft/s  %6.4f Mach\n",
					  stats.max_speed,
					  AltosConvert.meters_to_feet(stats.max_speed),
					  AltosConvert.meters_to_mach(stats.max_speed));
		if (stats.max_acceleration != AltosLib.MISSING) {
			System.out.printf("Max accel:   %6.0f m/s² %6.0f ft/s² %6.2f g\n",
					  stats.max_acceleration,
					  AltosConvert.meters_to_feet(stats.max_acceleration),
					  AltosConvert.meters_to_g(stats.max_acceleration));
		}
		if (stats.state_speed[Altos.ao_flight_drogue] != AltosLib.MISSING)
			System.out.printf("Drogue rate: %6.0f m/s  %6.0f ft/s\n",
					  stats.state_speed[Altos.ao_flight_drogue],
					  AltosConvert.meters_to_feet(stats.state_speed[Altos.ao_flight_drogue]));
		if (stats.state_speed[Altos.ao_flight_main] != AltosLib.MISSING)
			System.out.printf("Main rate:   %6.0f m/s  %6.0f ft/s\n",
					  stats.state_speed[Altos.ao_flight_main],
					  AltosConvert.meters_to_feet(stats.state_speed[Altos.ao_flight_main]));
		if (stats.landed_time != AltosLib.MISSING &&
		    stats.boost_time != AltosLib.MISSING &&
		    stats.landed_time > stats.boost_time)
			System.out.printf("Flight time: %6.0f s\n",
					  stats.landed_time -
					  stats.boost_time);
		System.out.printf("\n");
		return true;
	}

	static boolean process_oneline(File file) {
		AltosRecordSet set = record_set(file);
		if (set == null)
			return false;
		System.out.printf("%s", file.toString());
		AltosFlightSeries series = make_series(set);
		TestStats stats = new TestStats(series);
		if (stats.max_height != AltosLib.MISSING)
			System.out.printf(" height  %6.0f m", stats.max_height);
		if (stats.max_speed != AltosLib.MISSING)
			System.out.printf(" speed   %6.0f m/s", stats.max_speed);
		if (stats.state_enter_speed[AltosLib.ao_flight_drogue] != AltosLib.MISSING)
			System.out.printf(" drogue-deploy   %6.0f m/s", stats.state_enter_speed[AltosLib.ao_flight_drogue]);
		if (stats.max_acceleration != AltosLib.MISSING)
			System.out.printf(" accel   %6.0f m/s²", stats.max_acceleration);
		System.out.printf("\n");
		return true;
	}

	public static void help(int code) {
		System.out.printf("Usage: teststand [OPTION]... [FILE]...\n");
		System.out.printf("  Options:\n");
		System.out.printf("    --graph <filename>\t\tGraph a test\n");
		System.out.printf("    --summary <filename>\tText summary of a test\n");
		System.out.printf("    --oneline <filename>\tOne line summary of a test\n");
		System.out.printf("    --csv\t\t\tGenerate comma separated output for spreadsheets, etc\n");
		System.exit(code);
	}

	public static void main(final String[] args) {
		int	errors = 0;
		load_library(null);
		try {
			UIManager.setLookAndFeel(AltosUIPreferences.look_and_feel());
		} catch (Exception e) {
		}
		TestStand teststand = null;

		/* Handle batch-mode */
		if (args.length == 0) {
			teststand = new TestStand();
			java.util.List<AltosDevice> devices = AltosUSBDevice.list(Altos.product_basestation);
			if (devices != null)
				for (AltosDevice device : devices)
					teststand.telemetry_window(device);
		} else {
			int process = process_none;
			for (int i = 0; i < args.length; i++) {
				if (args[i].equals("--help"))
					help(0);
				else if (args[i].equals("--csv"))
					process = process_csv;
				else if (args[i].equals("--graph"))
					process = process_graph;
				else if (args[i].equals("--summary"))
					process = process_summary;
				else if (args[i].equals("--oneline"))
					process = process_oneline;
				else if (args[i].startsWith("--"))
					help(1);
				else {
					File file = new File(args[i]);
					switch (process) {
					case process_none:
						if (teststand == null)
							teststand = new TestStand();
					case process_graph:
						if (!process_graph(file))
							++errors;
						break;
					case process_csv:
						if (!process_csv(file))
							++errors;
						break;
					case process_summary:
						if (!process_summary(file))
							++errors;
						break;
					case process_oneline:
						if (!process_oneline(file))
							++errors;
						break;
					}
				}
			}
		}
		if (errors != 0)
			System.exit(errors);
	}
}
